Optimizations and fixes for @let declarations

crisbeto opened the pull request:
Includes the following optimizations/fixes for how `@let` declarations are handled:

### perf(compiler): reduce allocations for let declarations only used in the same view
We have some code that avoids `storeLet` calls for declarations only used in the same view, however we didn't previously remove the corresponding `declareLet` calls, because of the following case:

```
@let foo = something$ | async; <!-- First in the template -->
{{foo}}
```

Here we need a `TNode` (created by `declareLet`) in order for DI to work correctly. Since this is only required when using pipes, we can optimize away expressions that do not have pipes.


### fix(compiler): incorrectly handling let declarations inside i18n
The compiler wasn't handling `@let` declarations placed inside i18n blocks. The problem is that `assignI18nSlotDependencies` phase assigns the `target` of i18n ops much earlier than the `@let` optimization. If the `@let` ends up getting optimized because it isn't used in any child views, the pointer in the i18n instruction becomes invalid. This hadn't surfaced so far, because we didn't optimize `declareLet` ops, however once we do, we start hitting assertions that the optimized `declareLet` isn't used anywhere.

These changes resolve the issue by moving the i18n phases after the `@let` optimization.

crisbeto commented on packages/compiler-cli/test/compliance/test_cases/r3_view_compiler_let/let_multiple_optimization_template.js:
```diff
@@ -1,20 +1,19 @@
       $r3$.ɵɵtext(0);
-      $r3$.ɵɵdeclareLet(1)(2)(3)(4);
-      $r3$.ɵɵtext(5);
+      $r3$.ɵɵtext(1);
```
Note to self: we should also chain `text` instructions.

pkozlowski-opensource replied:
Would we have cases where 2 text nodes are side-by-side?

crisbeto replied:
I think the only case would be something like this where they were before/after a `@let` that was optimized away.

crisbeto committed:
perf(compiler): reduce allocations for let declarations only used in the same view

We have some code that avoids `storeLet` calls for declarations only used in the same view, however we didn't previously remove the corresponding `declareLet` calls, because of the following case:

```
@let foo = something$ | async; <!-- First in the template -->
{{foo}}
```

Here we need a `TNode` (created by `declareLet`) in order for DI to work correctly. Since this is only required when using pipes, we can optimize away expressions that do not have pipes.

crisbeto commented:
[Passing TGP](http://test/OCL:739135761:BASE:739980260:1742835916617:da47f5ed)

mmalerba commented on packages/compiler/src/template/pipeline/src/phases/store_let_optimization.ts:
```diff
@@ -29,17 +34,43 @@ export function optimizeStoreLet(job: CompilationJob): void {
+function hasPipe(root: ir.StoreLetExpr): boolean {
+  let result = false;
+
+  ir.transformExpressionsInExpression(
```
(no action required) kind of odd that there's know way to walk this thing without "transforming" it

crisbeto replied:
Agreed, but I think it's been like that since we switched to the pipeline.

mmalerba commented on packages/compiler/src/template/pipeline/src/phases/assign_i18n_slot_dependencies.ts:
```diff
@@ -67,7 +67,22 @@ export function assignI18nSlotDependencies(job: CompilationJob) {
+            target = updateOp.target;
+          } else if (
+            // Some expressions may consume slots as well (e.g. `storeLet`).
+            updateOp.kind === ir.OpKind.Statement ||
```
This assumes an op can only have one slotful expression, is that true?

crisbeto committed:
fix(compiler): incorrectly handling let declarations inside i18n

The compiler wasn't handling `@let` declarations placed inside i18n blocks. The problem is that `assignI18nSlotDependencies` phase assigns the `target` of i18n ops much earlier than the `@let` optimization. If the `@let` ends up getting optimized because it isn't used in any child views, the pointer in the i18n instruction becomes invalid. This hadn't surfaced so far, because we didn't optimize `declareLet` ops, however once we do, we start hitting assertions that the optimized `declareLet` isn't used anywhere.

These changes resolve the issue by moving the i18n phases after the `@let` optimization.

alxhub commented:
This PR was merged into the repository by commit 8f2874e86d8ed8c7b713713a9177697fe6d8f776.

The changes were merged into the following branches: main, 20.0.x
